src/algebra/modeq.lean

```diff
@@ -0,0 +1,212 @@
+
+alias modeq_neg ↔ modeq.of_neg' modeq.neg'
+
+@[simp] lemma zsmul_modeq_zsmul [no_zero_smul_divisors ℤ α] (hn : n ≠ 0) :
+  n • a ≡ n • b [PMOD (n • p)] ↔ a ≡ b [PMOD p] :=
+exists_congr $ λ m, by rw [←smul_sub, smul_comm, smul_right_inj hn]; apply_instance
+
+@[simp] lemma nsmul_modeq_nsmul [no_zero_smul_divisors ℕ α] {n : ℕ} (hn : n ≠ 0) :
+  n • a ≡ n • b [PMOD (n • p)] ↔ a ≡ b [PMOD p] :=
+exists_congr $ λ m, by rw [←smul_sub, smul_comm, smul_right_inj hn]; apply_instance
+
+alias zsmul_modeq_zsmul ↔ modeq.zsmul_cancel _
+alias nsmul_modeq_nsmul ↔ modeq.nsmul_cancel _
+
+lemma modeq_sub (a b : α) : a ≡ b [PMOD b - a] := ⟨1, (one_smul _ _).symm⟩
+
+@[simp] lemma modeq_zero : a ≡ b [PMOD 0] ↔ a = b := by simp [modeq, sub_eq_zero, eq_comm]
+
+@[simp] lemma zsmul_modeq_zero (n : ℤ) : n • p ≡ 0 [PMOD p] := ⟨-n, by simp⟩
+@[simp] lemma self_modeq_zero : p ≡ 0 [PMOD p] := ⟨-1, by simp⟩
+
+lemma add_zsmul_modeq : a + n • p ≡ a [PMOD p] := ⟨-n, by simp⟩
+lemma zsmul_add_modeq : n • p + a ≡ a [PMOD p] := ⟨-n, by simp⟩
+lemma add_nsmul_modeq {n : ℕ} : a + n • p ≡ a [PMOD p] := ⟨-n, by simp⟩
+lemma nsmul_add_modeq {n : ℕ} : n • p + a ≡ a [PMOD p] := ⟨-n, by simp⟩
+
+namespace modeq
+
+protected lemma of_zsmul : a ≡ b [PMOD (n • p)] → a ≡ b [PMOD p] :=
+λ ⟨m, hm⟩, ⟨m * n, by rwa [mul_smul]⟩
+
+protected lemma of_nsmul {n : ℕ} : a ≡ b [PMOD (n • p)] → a ≡ b [PMOD p] :=
+λ ⟨m, hm⟩, ⟨m * n, by rwa [mul_smul, coe_nat_zsmul]⟩
+
+protected lemma zsmul : a ≡ b [PMOD p] → n • a ≡ n • b [PMOD (n • p)] :=
+Exists.imp $ λ m hm, by rw [←smul_sub, hm, smul_comm]
+
+protected lemma nsmul {n : ℕ} : a ≡ b [PMOD p] → n • a ≡ n • b [PMOD (n • p)] :=
+Exists.imp $ λ m hm, by rw [←smul_sub, hm, smul_comm]
+
+@[simp] protected lemma add_iff_left :
+  a₁ ≡ b₁ [PMOD p] → (a₁ + a₂ ≡ b₁ + b₂ [PMOD p] ↔ a₂ ≡ b₂ [PMOD p]) :=
+λ ⟨m, hm⟩, (equiv.add_left m).symm.exists_congr_left.trans $
+  by simpa [add_sub_add_comm, hm, add_smul]
+
+@[simp] protected lemma add_iff_right :
+  a₂ ≡ b₂ [PMOD p] → (a₁ + a₂ ≡ b₁ + b₂ [PMOD p] ↔ a₁ ≡ b₁ [PMOD p]) :=
+λ ⟨m, hm⟩, (equiv.add_right m).symm.exists_congr_left.trans $
+  by simpa [add_sub_add_comm, hm, add_smul]
+
+@[simp] protected lemma sub_iff_left :
+  a₁ ≡ b₁ [PMOD p] → (a₁ - a₂ ≡ b₁ - b₂ [PMOD p] ↔ a₂ ≡ b₂ [PMOD p]) :=
+λ ⟨m, hm⟩, (equiv.sub_left m).symm.exists_congr_left.trans $
+  by simpa [sub_sub_sub_comm, hm, sub_smul]
+
+@[simp] protected lemma sub_iff_right :
+  a₂ ≡ b₂ [PMOD p] → (a₁ - a₂ ≡ b₁ - b₂ [PMOD p] ↔ a₁ ≡ b₁ [PMOD p]) :=
+λ ⟨m, hm⟩, (equiv.sub_right m).symm.exists_congr_left.trans $
+  by simpa [sub_sub_sub_comm, hm, sub_smul]
+
+alias modeq.add_iff_left ↔ add_left_cancel add
+alias modeq.add_iff_right ↔ add_right_cancel _
+alias modeq.sub_iff_left ↔ sub_left_cancel sub
+alias modeq.sub_iff_right ↔ sub_right_cancel _
+
+attribute [protected] add_left_cancel add_right_cancel add sub_left_cancel sub_right_cancel sub
+
+-- The following eight can already be proved by simp
```

Comment:
This comment seems useless to me; it easily can go out of sync, and the below lemmas are conditional simp lemmas anyway?

Reply:
Well, I had them as simp at some point and it caused a timeout in simp-nf, but sure I can remove the comment.